seps/sep-0002.md

```diff
@@ -0,0 +1,100 @@
+
+Build systems can be specified in specs using a key-value pair:
+```
+hdf5 build_system=cmake
+```
+__which requires the keyword `build_system` to be reserved by Spack__ (similarly to what is done for `dev_path`). It is interesting to note that modeling build-systems this way make them similar to signle-valued variants with the only difference that the set of allowed values is "dynamic" and subject to constraints.
+
+__Each package needs to have one and only one build-system__.
+
+### Overriding build-system phases
+
+Some packages may need to override phases of the build-systems they use and there must be a way to specify which phase for which build-system is overridden. What we propose to use here is a decorator to distinguish among multiple build-systems:
+```python
+class Hdf5(Package):
+    build_system('cmake', when='@X.Y:')
+    build_system('autotools')
+
+    @cmake
+    def build(self, spec, prefix):
+         pass
+		
+    @autotools
+    def build(self, spec, prefix):
+         pass
+```
+Using a decorator permits seamlessly to override phases with the same name.
+
+## Implementation at a high-level
+
+Currently the build-system specific methods are grouped in base classes and each package uses them via inheritance. There's a fair amount of machinery with metaclasses that is used to have dynamic phases for each build-system.
+
+To support multiple build-systems we should probably add, as it happens in other directives, a dictionary of builder objects:
+```python
+hdf5.build_system = {
+    'cmake': cmake_builder_obj,
+    'autotools': autotools_builder_obj
+}
+```
+
+Each builder object will accept a _package object tied to a concrete spec_ as argument and contain all the methods that are relevant for building it. 
+
+The main idea is to:
+1. Move things like `phases` and other similar methods into a class of their own
+2. Use a different object for each package so that the package can override these methods if need be 
+3. Have decorators that can monkey patch the object to customize the install phases
+
+Dipatching to the correct build system object happens once we have a concrete spec associated with the package (since the spec coveys information about the build system, we know which one to select).
+
+### Backward compatibility
+
+Most of the packages are currently fine with a single build-system and they inherit almost all of the related methods from a base class:
+
+```python
+class OpenJpeg(CmakePackage):
+	pass
+```
+
+To be backward compatible, if we push the methods and attributes down to a "builder" class, we can maintain the directives in the same base class:
+```python
+class CMakePackage(PackageBase):
+	# Can be overridden if conditional in derived classes
+	build_system('cmake')
+	depends_on('cmake', type='build')
```

Comment:
If I write a package that supports both CMake and Autotools, will I have to manually add:
```python
depends_on('cmake', when='build_system=cmake', type='build')
```
or can that be included in `CMakeBuildSystem`? Is the `build_system=cmake` even available in the spec?

Reply:
> Is the build_system=cmake even available in the spec?

Yes, that is available in the spec. There are legitimate use cases where you want to use it in recipes too, e.g. for bootstrapping tools. Think for instance `foo` can build with `cmake` and `makefile` and is a `cmake` dependency. In that recipe you probably want to specify:
```python
depends_on('foo build_system=makefile')
```
to avoid circular dependencies.

> If I write a package that supports both CMake and Autotools, will I have to manually add: [ ... ]

I think that should be implied and something similar needs to be included only if there are additional constraints to be met e.g. on the version of `cmake`.